### PR TITLE
Kotlin migration DatabaseErrorDialog

### PR DESCRIPTION
**Pull request by Arnold2381**
Kotlin migration DatabaseErrorDialog

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 1]
```diff
@@ -14,439 +14,428 @@
...
-                        .content(getMessage())
-                        .iconAttr(R.attr.dialogErrorIcon)
-                        .positiveText(R.string.error_handling_options)
-                        .negativeText(R.string.close)
-                        .onPositive((inner_dialog, which) -> ((DeckPicker) getActivity())
-                                .showDatabaseErrorDialog(DIALOG_ERROR_HANDLING))
-                        .onNegative((inner_dialog, which) -> exit())
-                        .show();
+                builder.cancelable(false)
+                    .content(message!!)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:59]: !

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 2]
```diff
@@ -14,439 +14,428 @@
...
-                            dismissAllDialogFragments();
-                        })
-                        .onNeutral((inner_dialog, which) -> exit())
-                        .show();
-                dialog.getCustomView().findViewById(R.id.md_buttonDefaultNegative).setEnabled(
-                        ((DeckPicker) getActivity()).hasErrorFiles());
-                return dialog;
+                val dialog = builder
+                    .cancelable(false)
+                    .content(message!!)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:76]: !

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 3]
```diff
@@ -14,439 +14,428 @@
...
-                return builder.content(getMessage())
-                        .iconAttr(R.attr.dialogErrorIcon)
-                        .positiveText(R.string.dialog_positive_repair)
-                        .negativeText(R.string.dialog_cancel)
-                        .onPositive((inner_dialog, which) -> {
-                            ((DeckPicker) getActivity()).repairCollection();
-                            dismissAllDialogFragments();
-                        })
-                        .show();
+                builder.content(message!!)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:168]: !

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 4]
```diff
@@ -14,439 +14,428 @@
...
-                            ch.closeCollection(false, "DatabaseErrorDialog: Before Create New Collection");
-                            String path1 = CollectionHelper.getCollectionPath(getActivity());
-                            if (BackupManager.moveDatabaseToBrokenFolder(path1, false, time)) {
-                                ((DeckPicker) getActivity()).restartActivity();
-                            } else {
-                                ((DeckPicker) getActivity()).showDatabaseErrorDialog(DIALOG_LOAD_FAILED);
-                            }
-                        })
-                        .show();
+                builder.content(message!!)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:246]: !

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 5]
```diff
@@ -14,439 +14,428 @@
...
                 // Confirmation dialog for database check
-                return builder.content(getMessage())
-                        .positiveText(R.string.dialog_ok)
-                        .negativeText(R.string.dialog_cancel)
-                        .onPositive((inner_dialog, which) -> {
-                            ((DeckPicker) getActivity()).integrityCheck();
-                            dismissAllDialogFragments();
-                        })
-                        .show();
+                builder.content(message!!)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:265]: !

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 6]
```diff
@@ -14,439 +14,428 @@
...
+            DIALOG_CONFIRM_RESTORE_BACKUP -> {
+
                 // Confirmation dialog for backup restore
-                return builder.content(getMessage())
-                        .positiveText(R.string.dialog_continue)
-                        .negativeText(R.string.dialog_cancel)
-                        .onPositive((inner_dialog, which) -> ((DeckPicker) getActivity())
-                                .showDatabaseErrorDialog(DIALOG_RESTORE_BACKUP))
-                        .show();
+                builder.content(message!!)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:277]: !

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 7]
```diff
@@ -14,439 +14,428 @@
...
                 // Allow user to do a full-sync from the server
-                return builder.content(getMessage())
-                        .positiveText(R.string.dialog_positive_overwrite)
-                        .negativeText(R.string.dialog_cancel)
-                        .onPositive((inner_dialog, which) -> {
-                            ((DeckPicker) getActivity()).sync(Connection.ConflictResolution.FULL_DOWNLOAD);
-                            dismissAllDialogFragments();
-                        })
-                        .show();
+                builder.content(message!!)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:289]: !

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 8]
```diff
@@ -14,439 +14,428 @@
...
-                //If the database is locked, all we can do is ask the user to exit.
-                return builder.content(getMessage())
-                        .positiveText(R.string.close)
-                        .cancelable(false)
-                        .onPositive((inner_dialog, which) -> exit())
-                        .show();
+            DIALOG_DB_LOCKED -> {
+
+                // If the database is locked, all we can do is ask the user to exit.
+                builder.content(message!!)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:301]: !

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 9]
```diff
@@ -14,439 +14,428 @@
...
-                        })
-                        .show();
+            INCOMPATIBLE_DB_VERSION -> {
+                val values: MutableList<Int> = ArrayList(2)
+                val options = arrayOf<CharSequence>(makeBold(res.getString(R.string.backup_restore)), makeBold(res.getString(R.string.backup_full_sync_from_server)))
+                values.add(0)
+                values.add(1)
+                builder
+                    .cancelable(false)
+                    .content(message!!)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:314]: !

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 10]
```diff
@@ -14,439 +14,428 @@
...
+            DIALOG_CONFIRM_RESTORE_BACKUP -> res().getString(R.string.restore_backup_title)
+            DIALOG_FULL_SYNC_FROM_SERVER -> res().getString(R.string.backup_full_sync_from_server)
+            DIALOG_DB_LOCKED -> res().getString(R.string.database_locked_title)
+            INCOMPATIBLE_DB_VERSION -> res().getString(R.string.incompatible_database_version_title)
+            DIALOG_DB_ERROR -> res().getString(R.string.answering_error_title)
+            else -> res().getString(R.string.answering_error_title)
         }
+
+    override fun getNotificationMessage(): String {
+        return message!!
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:386]: Nullable

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 11]
```diff
@@ -14,439 +14,428 @@
...
-                return res().getString(R.string.backup_full_sync_from_server);
-            case DIALOG_DB_LOCKED:
-                return res().getString(R.string.database_locked_title);
-            case INCOMPATIBLE_DB_VERSION:
-                return res().getString(R.string.incompatible_database_version_title);
-            case DIALOG_DB_ERROR:
-            default:
-                return res().getString(R.string.answering_error_title);
-        }        
+    override fun getNotificationTitle(): String {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:389]: Make nullable

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Few issues, nothing major, we likely want to aim to get the base class converted so the overloads aren't erroneously returning null

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt (hunk 11 above)
Comment (Arnold2381) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:389]: same it always return non null value

Comment (Arnold2381) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:390]:
```
override fun getNotificationTitle(): String? {
        return res().getString(R.string.answering_error_title)
    }

```

Comment (Arnold2381) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:390]: @david-allison-1 ?

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:390]: Sorry, that's fine

**Review by david-allison-1 — APPROVED** [review]
LGTM

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 12]
```diff
@@ -14,439 +14,429 @@
...
+        var sqliteInstalled = false
         try {
-            sqliteInstalled = Runtime.getRuntime().exec("sqlite3 --version").waitFor() == 0;
-        } catch (IOException | InterruptedException e) {
-            Timber.w(e);
+            sqliteInstalled = Runtime.getRuntime().exec("sqlite3 --version").waitFor() == 0
+        } catch (e: IOException) {
+            Timber.w(e)
+        } catch (e: InterruptedException) {
+            Timber.w(e)
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:52]: Interesting. Multi-catch not supported, vote here if you like https://youtrack.jetbrains.com/issue/KT-7128

**Review by mikehardy — APPROVED** [review]
Alright!

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt (hunk 12 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:52]: I don't see it happening soon, Exceptions are a rare case where Java allows sum types. I'd love to see it, but it's a paradigm shift

**Pull request merged**